zio-interop-cats/shared/src/main/scala/zio/stream/interop/FS2StreamSyntax.scala

```diff
@@ -40,37 +40,65 @@ final class FS2RIOStreamSyntax[R, A](private val stream: Stream[RIO[R, _], A]) {
-            .eval(queue.offer(Take.end)))
-            .handleErrorWith(e => fs2.Stream.eval(queue.offer(Take.fail(e))).drain)
-            .compile[RIO[R, _], RIO[R, _], Any]
-            .resource
-            .drain
-            .toManagedZIO
-            .fork
-      } yield ZStream.fromQueue(queue).flattenTake
-    }.flatten
-
-  private def toZStreamChunk[R1 <: R](queueSize: Int): ZStream[R1, Throwable, A] =
-    ZStream.managed {
-      for {
-        queue <- Queue.bounded[Take[Throwable, A]](queueSize).toManaged(_.shutdown)
-        _     <- {
-          stream
-            .chunkLimit(queueSize)
-            .evalTap(a => queue.offer(Take.chunk(zio.Chunk.fromIterable(a.toList))))
-            .unchunk ++ fs2.Stream.eval(queue.offer(Take.end))
-        }.handleErrorWith(e => fs2.Stream.eval(queue.offer(Take.fail(e))).drain)
-          .compile[RIO[R, _], RIO[R, _], Any]
-          .resource
+  def toZStreamChunk[R1 <: R](queueSize: Int = 16): ZStream[R1, Throwable, A] = {
+    def integrate(stream: fs2.Stream[RIO[R, _], A], zQueue: Queue[Take[Throwable, A]]): fs2.Pull[RIO[R, _], A, Unit] =
+      stream.pull.uncons.flatMap {
+        case None =>
+          fs2.Pull.eval(zQueue.offer(Take.end)) >> fs2.Pull.done
+
+        case Some((fs2Chunk: fs2.Chunk[A], stream)) =>
+          val offer = zQueue.offer(Take.chunk(toZioChunk(fs2Chunk)))
+          fs2.Pull.eval(offer) >> integrate(stream, zQueue)
+      }
+
+    convert(queueSize, integrate)
+  }
+
+  private def toZStreamSingle[R1 <: R]: ZStream[R1, Throwable, A] = {
+    def integrate(stream: fs2.Stream[RIO[R, _], A], zQueue: Queue[Take[Throwable, A]]): fs2.Pull[RIO[R, _], A, Unit] =
+      stream.pull.uncons.flatMap {
+        case None =>
+          fs2.Pull.eval(zQueue.offer(Take.end)) >> fs2.Pull.done
+
+        case Some((elements, stream)) =>
+          val offer = zQueue.offerAll(toZioChunk(elements).map(Take.single))
+          fs2.Pull.eval(offer) >> integrate(stream, zQueue)
+      }
+
+    convert(1, integrate)
+  }
+
+  private def convert(
+    queueSize: Int,
+    integrate: (fs2.Stream[RIO[R, _], A], Queue[Take[Throwable, A]]) => fs2.Pull[RIO[R, _], A, Unit]
+  ) =
+    ZStream.fromEffect(Queue.bounded[Take[Throwable, A]](queueSize)).flatMap { q =>
+      val toQueue = ZStream.fromEffect {
+        integrate(stream, q).stream
+          .handleErrorWith(e => fs2.Stream.eval(q.offer(Take.fail(e))))
+          .compile
           .drain
-          .toManagedZIO
-          .fork
-      } yield ZStream.fromQueue(queue).flattenTake
-    }.flatten
+      }
+
+      val fromQueue = ZStream.fromQueue(q).flattenTake
+      fromQueue.drainFork(toQueue)
+    }
+
+  private def toZioChunk(in: fs2.Chunk[A]): Chunk[A] =
+    in match {
+      case fs2.Chunk.ArraySlice(values, _, _) =>
+        Chunk.fromArray(values)
+
+      case buffer: fs2.Chunk.Buffer[_, _, _] =>
+        Chunk.fromIterator(buffer.iterator)
```

Comment:
The documentation for `buffer.iterator` mentions that the returned iterator is not thread-safe so I'm wondering if this is still safe to use?

Reply:
Yes this should be fine because there is only one thread interacting with the iterator, the one running `Chunk.fromIterator`, and it will just read the iterator and place its contents into an immutable data structure that will return and will be safe for concurrent access.